---
title: 'RNA-Seq workflow: Thyroid analysis within 3 type of infiltration'
author: 'Ana Isabel del Val'
subtitle: '`r params$subtitulo`'
date: '`r format(Sys.Date(),"%e de %B, %Y")`' 
# date: \today  (solo para pdf)
output:
  pdf_document:
    keep_tex: true
    number_sections: no
    toc: true
    toc_depth: 3
  #html_document:
   # toc: yes
    #toc_depth: 3
#nocite: | 
 # @github
 # @source
params:
  subtitulo: Análisis de datos ómicos PEC2
bibliography: PEC2.bib
geometry: margin=2cm
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
options(width=90)
```

```{r include=FALSE}
if(!require(BiocManager)) install.packages("BiocManager")
if(!require(airway)) BiocManager::install("airway")
if(!require(Rsamtools)) BiocManager::install("Rsamtools")
if(!require(GenomicFeatures)) BiocManager::install("GenomicFeatures")
if(!require(DESeq2)) BiocManager::install("DESeq2")
if(!require(apeglm)) BiocManager::install("apeglm")
if(!require(BiocParallel)) BiocManager::install("BiocParallel")
if(!require(genefilter)) BiocManager::install("genefilter")
if(!require(org.Hs.eg.db)) BiocManager::install("org.Hs.eg.db")
if(!require(AnnotationDbi)) BiocManager::install("AnnotationDbi")
if(!require(ReportingTools)) BiocManager::install("ReportingTools")
if(!require(RUVSeq)) BiocManager::install("RUVSeq")
if(!require(sva)) BiocManager::install("sva")
if(!require(Gviz)) BiocManager::install("Gviz")
if(!require(magrittr)) install.packages("magrittr", dep=TRUE)
if(!require(dplyr)) install.packages("dplyr", dep=TRUE)
if(!require(ggplot2)) install.packages("ggplot2", dep=TRUE)
if(!require(pheatmap)) install.packages("pheatmap", dep=TRUE)
if(!require(RColorBrewer)) install.packages("RColorBrewer", dep=TRUE)
if(!require(ggbeeswarm)) install.packages("ggbeeswarm", dep=TRUE)
if(!require(dplyr)) install.packages("dplyr", dep=TRUE)
BiocManager::install("org.Hs.eg.db")
```

## Abstract
The dataset for the exercise comes from the repository GTEx and has been provided by UOC. This repository contains data of multiple types out of 54 tissues. We focus on RNA-seq expression data of a thyroid analysis, where 3 type of infiltration is compared in 292 samples.

The distribution of the samples is the following:
- Not infiltrated tissues (NIT): 236 samples
- Small focal infiltrates (SFI): 42 samples
- Extensive lymphoid infiltrates (ELI): 14 samples

We already have the preprocessed data in the count matrix, _counts.csv_ file.

The distribution of samples are stored in _targets.csv_, where we select 30 samples (10 of each type). These 30 samples will subset the columns of the count matrix, having rows of targets and columns of counts in the same order.

This code is already sync with __*[github repository](https://github.com/AnadelVal/RNA-seq_Thyroid)*__, which is public. 

__*[This is the Bioconductor help](https://www.bioconductor.org/help/course-materials/2017/CSAMA/labs/2-tuesday/lab-03-rnaseq/rnaseqGene_CSAMA2017.html#experimental-data)*__.

## Objectives
It consists on the differential expression analysis, comparing 3 type of Group samples:

* NIT-SFI
* NIT-ELI
* SFI-ELI

## Materials and methods

### Data source and experiment design

We count on 292 samples of 3 types out of 54 tissues. We select 30 samples out of 292, equally distributed by group.

* 10 samples of group NIT
* 10 samples of group ELI
* 10 samples of group SFI

Data has 2 molecular data type:

* RNA-Seq (NGS)
* Allele-Specific Expression

If we filter by RNA-Seq (NGS) only, there are only 8 samples of ELI group. Consequently, we have consider the two molecular data types to get the random samples.

###  Methods and Tools

R and Bioconductor have been the tools selected to follow the RNA-seq pipeline:

* Data gathering
* Data preprocessing 
* Data filtering and transformations
* Differential expression analysis
* Results comparisons
    + NIT_SFI
    + NIT_ELI
    + SFI_ELI
* Results annotations
    + NIT_SFI
    + NIT_ELI
    + SFI_ELI
* Remove unwanted variations

## Results

The data preparation is crutial for obtaining good results. These are the highlights:

* The order of rows in targets is the same as the order of columns in counts.

* Remove ENSEMBLE version from genes.

* Standardize naming convention for samples (be careful not to change "-" by "." in columns of counts when loading data). 

* The seed is fixed to "1234"to ensure results reproducibility.

```{r include=FALSE}
targets <- read.csv("targets.csv", sep=",", row.names=NULL)
head(targets)
str(targets$Group)
str(targets$Grupo_analisis)
str(targets)
set.seed(1234)
targets_NIT <- targets[sample(which(targets$Group == "NIT" ),10),]
targets_SFI <- targets[sample(which(targets$Group == "SFI" ),10),]
targets_ELI <- targets[sample(which(targets$Group == "ELI" ),10),]
targets_30 <- bind_rows(targets_NIT, targets_SFI, targets_ELI)

counts <- read.csv("counts.csv", sep=";", row.names=1, check.names=FALSE)
selection <- names(counts) %in% as.character(targets_30$Sample_Name)
counts_30 <- counts[selection]
str(counts_30)

rownames(targets_30) <- targets_30[,3]
counts_30 <- counts_30[,order(names(counts_30))]
targets_30 <- targets_30[order(row.names(targets_30)),]
row.names(counts_30) <- gsub("\\..*", "", row.names(counts_30), fixed = FALSE)

ddsMatrix <- DESeqDataSetFromMatrix(countData = counts_30,
                                 colData = targets_30,
                                 design = ~ Group)
```

In the definition of DESeqDataSet object, we specify the variable "Group" to test for its effect in the experiment.
```{r}
ddsMatrix
```

Our count matrix with our DESeqDataSet contains many rows with only zeros, and additionally many rows with only a few fragments total. In order to reduce the size of the object, and to increase the speed of our functions, we can remove the rows that have no or nearly no information about the amount of gene expression.
Here we apply the most minimal filtering rule: removing rows of the DESeqDataSet that have no counts, or only a single count across all samples.

```{r}
ddsMatrix <- ddsMatrix[ rowSums(counts(ddsMatrix)) > 1, ]
nrow(ddsMatrix)
```

DESeq2 offers two transformations for count data that stabilize the variance across the mean: the regularized-logarithm transformation or rlog and the variance stabilizing transformation (VST) for negative binomial data with a dispersion-mean trend. In this case we choose rlog because the number of samples is 30, and rlog tends to work well with small datasets.
```{r include=FALSE}
rldM <- rlog(ddsMatrix, blind = FALSE)
head(assay(rldM), 3)
```

Now we assess the sample distances, which is the overall similarity between samples.
The distance function expect samples in rows and genes in columns, that's why we traspose the matrix coming from rlog transformation. We use the heatmap plot to see the sample distances.
```{r}
sampleDists <- dist(t(assay(rldM)))
sampleDistMatrix <- as.matrix(sampleDists)
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

When performing the differential expression analysis and later comparisons between groups, we consider a fraction of 10% false positives acceptable so we can consider all genes with an adjusted p value below 10% = 0.1 as significant.

```{r include=FALSE}
ddsMatrix <- DESeq(ddsMatrix, parallel =TRUE)
### Comparison 1
resNIT_SFI <- results(ddsMatrix, contrast=c("Group","NIT","SFI")) 
resNIT_SFI
summary(resNIT_SFI)

###Comparison 2
resNIT_ELI <- results(ddsMatrix, contrast=c("Group","NIT","ELI")) 
resNIT_ELI
summary(resNIT_ELI)

###Comparison 3
resSFI_ELI <- results(ddsMatrix, contrast=c("Group","SFI","ELI")) 
resSFI_ELI
summary(resSFI_ELI)
```

These are the significant genes for the 3 comparisons, corresponding to NIT-SFI, NIT-ELI and SFI-ELI, respectively. There is a huge difference between the first comparison and the other 2.
```{r}
sum(resNIT_SFI$padj < 0.1, na.rm=TRUE)
sum(resNIT_ELI$padj < 0.1, na.rm=TRUE)
sum(resSFI_ELI$padj < 0.1, na.rm=TRUE)
```

We proceed with the results annotation. The columns $symbol and $entrez could lead us to NA values. The most justified reason is that there could be transcripts not associated with any specific gene.
```{r include=FALSE}
##Comparison 1
resNIT_SFI$symbol <- mapIds(org.Hs.eg.db,
                            keys=row.names(resNIT_SFI),
                            column="SYMBOL",
                            keytype="ENSEMBL",
                            multiVals="first")
resNIT_SFI$entrez <- mapIds(org.Hs.eg.db,
                            keys=row.names(resNIT_SFI),
                            column="ENTREZID",
                            keytype="ENSEMBL",
                            multiVals="first")

resNIT_SFIOrdered <- resNIT_SFI[order(resNIT_SFI$pvalue),]
resNIT_SFIOrdered <- as.data.frame(resNIT_SFIOrdered)
write.csv(resNIT_SFIOrdered, file = "resultsNIT_SFI.csv")

##Comparison 2
resNIT_ELI$symbol <- mapIds(org.Hs.eg.db,
                            keys=row.names(resNIT_ELI),
                            column="SYMBOL",
                            keytype="ENSEMBL",
                            multiVals="first")
resNIT_ELI$entrez <- mapIds(org.Hs.eg.db,
                            keys=row.names(resNIT_ELI),
                            column="ENTREZID",
                            keytype="ENSEMBL",
                            multiVals="first")

resNIT_ELIOrdered <- resNIT_ELI[order(resNIT_ELI$pvalue),]
resNIT_ELIOrdered <- as.data.frame(resNIT_ELIOrdered)
write.csv(resNIT_ELIOrdered, file = "resultsNIT_ELI.csv")

###Comparison 3
library("AnnotationDbi")
resSFI_ELI$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(resSFI_ELI),
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
resSFI_ELI$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(resSFI_ELI),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

resSFI_ELIOrdered <- resSFI_ELI[order(resSFI_ELI$pvalue),]
resSFI_ELIOrdered <- as.data.frame(resSFI_ELIOrdered)
write.csv(resSFI_ELIOrdered, file = "resultsSFI_ELI.csv")
```

```{r}
head(resNIT_SFIOrdered)
head(resNIT_ELIOrdered)
head(resSFI_ELIOrdered)
```

It's time to remove hidden batch effects causing unwanted variations. This will detect the source of variation correlated with Group.
```{r include=FALSE}
dat  <- counts(ddsMatrix, normalized = TRUE)
idx  <- rowMeans(dat) > 1
dat  <- dat[idx, ]
mod  <- model.matrix(~ Group, colData(ddsMatrix))
mod0 <- model.matrix(~   1, colData(ddsMatrix))
svseq <- svaseq(dat, mod, mod0, n.sv = 2)
svseq$sv
```

```{r}
par(mfrow = c(2, 1), mar = c(3,5,3,1))
for (i in 1:2) {
  stripchart(svseq$sv[, i] ~ ddsMatrix$Group, vertical = TRUE, main = paste0("SV", i))
  abline(h = 0)
}
```
We could then produce results by running DESeq with the new design, incorporating the surrogate variables.

## Discussion

I have found a limitation with the first comparison: NIT_SFI. When getting the significant genes, we only get around 1.5% from Comparison 1 with respecto to Groups 2 and 3. So the results for the first comparison could be less accurate.

We could have plotted a lot of visualizations, but as we discussed in the class forum, there is no point in representing the same with different methods. This way, only heatmap has been chosen for sample distances.